Adding optimization with DE and a corresponding example + tests

phoebe-p opened the pull request:
This adds a module, 'optimization', which contains differential evolution functions, slightly modified from the package Yabox (credit to Pablo R. Mier given in the code source). It still uses the internal functions of yabox, just slightly changes the output to include more details of the optimization. It also adds yabox to the requirements (I'm not sure I did this correctly or if it also needs to be added somewhere else, but the tests run successfully). It also adds an example of using DE to optimize a four-junction device optically and electrically.

dalonsoa commented on setup.py:
```diff
@@ -73,7 +73,7 @@ def gen_data_files(*dirs):
     "pyyaml",
 ]
-tests_require = ["pytest", "pytest-cov", "pytest-mock"]
+tests_require = ["pytest", "pytest-cov", "pytest-mock", "yabox"]
```
Well, this is not really a requirement only needed in the tests, but actually to run the optmizer, so it shuold be in either in the `install_requires` above or as a subtype of `extras_require`:
```python
extras_require = {"dev": tests_require + ["Sphinx", "pre-commit"], "opt": ["yabox"]}
```

dalonsoa commented on examples/differential_evolution_4Jcell.py:
```diff
@@ -0,0 +1,484 @@
+# try optimizing an anti-reflection coating
```
This example, more than an example, is a full-featured tutorial. It's absolutely fine - and increadibly useful as a real-world application -, but it would be useful to have some sort of minimum working example showing how the optimizar is used. Something simple of the type "Adjust the bandgap of a 1J solar cell for best performance under the AM1.5g spectrum" or something like that that can be done in a handful of lines of code.

dalonsoa commented on solcore/optimization/differential_evolution.py:
```diff
@@ -0,0 +1,158 @@
+"""
```
I've just realized that you have copied from the original repo the whole classes that you want to replace, and then tweak those parts that you don't like. I think that a better approach will be to create subclasses of the original classes and then override the relevant methods/attributes you want. Methods and atributes not overriden will be inherited from the original implementation.

Something like:
```python
class MyClass(ClassToReplace):
    def method_to_replace(self, p1, p2):
        "your implementation of the method"
        return whatever_you_like
```

Here `MyClass` will be identical to `ClassToReplace` except for the `method_to_replace`.

This way, it becomes clear what are the differences with the original code, makes your own code way shorter and, in summary, easier to debug.

phoebe-p replied:
good idea, thank you!

phoebe-p commented:
Added another example (much faster, takes a couple of seconds to run for me) and changed the requirements and how the DE and PDE methods are imported/amended from yabox.
